experimental/Schemes/WeilDivisor.jl

```diff
@@ -0,0 +1,352 @@
+export WeilDivisor
+export scheme_type, ideal_sheaf_type, coefficient_ring_type, coefficient_type
+export scheme, components, coefficient_dict, coefficient_ring
+export in_linear_system
+
+export LinearSystem 
+export divisor, find_subsystem
+
+@Markdown.doc """
+    WeilDivisor
+
+A Weil divisor on a covered scheme ``X``; 
+stored as a formal linear combination over 
+some ring ``R`` of ideal sheaves on ``X``.
+"""
+@attributes mutable struct WeilDivisor{
+    CoveredSchemeType<:AbsCoveredScheme, 
+    CoefficientRingType<:AbstractAlgebra.Ring, 
+    CoefficientRingElemType<:AbstractAlgebra.RingElem
+   }
+  X::CoveredSchemeType # the parent
+  R::CoefficientRingType # the ring of coefficients
+  coefficients::IdDict{IdealSheaf, CoefficientRingElemType} # the formal linear combination
+
+  function WeilDivisor(
+      X::AbsCoveredScheme,
+      R::CoefficientRingType, 
+      coefficients::IdDict{<:IdealSheaf, CoefficientRingElemType};
+      check::Bool=true
+    ) where {CoefficientRingType, CoefficientRingElemType}
+    # TODO: Do we want to require that the different effective divisors 
+    # have the same underlying covering? Probably not.
+    if check
+      for D in keys(coefficients)
+        space(D) === X || error("component of divisor does not lay in the given scheme")
+        parent(coefficients[D]) === R || error("coefficients do not lay in the given ring")
+      end
+    end
+    return new{typeof(X), CoefficientRingType, CoefficientRingElemType}(X, R, coefficients)
+  end
+end
+
+### type getters 
+scheme_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = S
+scheme_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = S
+coefficient_ring_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = U
+coefficient_ring_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = U
+coefficient_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = V
+coefficient_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = V
+
+### getter methods
+@Markdown.doc """
+    scheme(D::WeilDivisor)
+
+Return the `CoveredScheme` ``X`` on which `D` is defined.
+"""
+scheme(D::WeilDivisor) = D.X
+
+getindex(D::WeilDivisor, I::IdealSheaf) = (D.coefficients)[I]
+
+@Markdown.doc """
+    components(D::WeilDivisor)
+
+Return the irreducible components ``Eⱼ`` of the divisor 
+``D = Σⱼ aⱼ ⋅ Eⱼ``.
+"""
+components(D::WeilDivisor) = [ Z for Z in keys(D.coefficients)]
```

Comment:
I think hartshorne calls this the support of `D`.

Reply:
I did not find 'support' for this in the book of Hartshorne -- but maybe I simply did not find the right spot in the book.

Reply:
He does not really give the coefficients a formal name in a definition. Ctr+F gives a single proof where he calls it the support. How do other authors call it?

Reply:
Vakil defines the support of $D$ as the set 
$\bigcup_{n_Y \neq 0} Y$.